File: app/src/main/java/it/niedermann/nextcloud/deck/persistence/sync/adapters/db/dao/JoinCardWithUserDao.java
package it.niedermann.nextcloud.deck.persistence.sync.adapters.db.dao;

import androidx.room.Dao;
import androidx.room.Query;

import java.util.List;

import it.niedermann.nextcloud.deck.model.JoinCardWithUser;

@Dao
public interface JoinCardWithUserDao extends GenericDao<JoinCardWithUser> {
    @Query("DELETE FROM joincardwithuser WHERE cardId = :localId and status=1") // 1 = UP_TO_DATE
    void deleteByCardId(long localId);

    @Query("Update joincardwithuser set status = :status WHERE cardId = :localCardId and userId = :localUserId")
    void setDbStatus(long localCardId, long localUserId, int status);

    @Query("DELETE FROM joincardwithuser WHERE cardId = :localCardId and userId = :localUserId")
    void deleteByCardIdAndUserIdPhysically(long localCardId, long localUserId);

    @Query("select * FROM joincardwithuser WHERE cardId = :localCardId and userId = :localUserId")
    JoinCardWithUser getJoin(Long localUserId, Long localCardId);

    @Query("select u.localId as userId, c.id as cardId, j.status from joincardwithuser j " +
            "inner join card c on j.cardId = c.localId " +
            "inner join user u on j.userId = u.localId " +
            "WHERE j.status <> 1") // not UP_TO_DATE
    List<JoinCardWithUser> getChangedJoinsWithRemoteIDs();

    @Query("select u.localId as userId, c.id as cardId, j.status from joincardwithuser j " +
            "inner join card c on j.cardId = c.localId " +
            "inner join user u on j.userId = u.localId " +
            "WHERE c.stackId = :localStackId " +
            "AND j.status <> 1") // not UP_TO_DATE
    List<JoinCardWithUser> getChangedJoinsWithRemoteIDsForStack(Long localStackId);

    @Query("delete from joincardwithuser " +
            "where cardId = (select c.localId from card c where c.accountId = :accountId and c.id = :remoteCardId) " +
            "and userId = (select u.localId from user u where u.accountId = :accountId and u.uid = :userUid)")
    void deleteJoinedUserForCardPhysicallyByRemoteIDs(Long accountId, Long remoteCardId, String userUid);

    @Query("select userId from joincardwithuser WHERE cardId = :localCardId and userId IN (:assignedUserIDs) and status <> 3") // not LOCAL_DELETED
    List<Long> filterDeleted(long localCardId, List<Long> assignedUserIDs);
}

File: data/local/src/main/java/it/niedermann/nextcloud/deck/data/local/dao/JoinCardWithUserDao.kt
package it.niedermann.nextcloud.deck.data.local.dao

import androidx.room3.Dao
import androidx.room3.Query
import it.niedermann.nextcloud.deck.data.local.entity.JoinCardWithUserEntity
import java.util.concurrent.CompletableFuture

@Dao
interface JoinCardWithUserDao : GenericDao<JoinCardWithUserEntity> {
    @Query("DELETE FROM JoinCardWithUser WHERE cardId = :cardId")
    fun deleteByCardId(cardId: Long): CompletableFuture<Void?>

    @Query("DELETE FROM JoinCardWithUser WHERE cardId = :cardId AND userId = :userId")
    fun deleteByCardIdAndUserId(cardId: Long, userId: Long): CompletableFuture<Void?>

    @Query("SELECT * FROM JoinCardWithUser WHERE cardId = :cardId AND status != 3")
    fun getActiveJoinsByCardId(cardId: Long): CompletableFuture<List<JoinCardWithUserEntity>>

    @Query("SELECT * FROM JoinCardWithUser WHERE cardId = :cardId")
    fun getJoinsByCardId(cardId: Long): CompletableFuture<List<JoinCardWithUserEntity>>

    @Query("SELECT * FROM JoinCardWithUser WHERE cardId = :cardId AND userId = :userId")
    fun getJoin(cardId: Long, userId: Long): CompletableFuture<JoinCardWithUserEntity?>

    @Query("SELECT JoinCardWithUser.* FROM JoinCardWithUser INNER JOIN Card ON JoinCardWithUser.cardId = Card.localId WHERE Card.accountId = :accountId AND JoinCardWithUser.status != 1")
    fun getChangedJoinsForAccount(accountId: Long): CompletableFuture<List<JoinCardWithUserEntity>>

    @Query("UPDATE JoinCardWithUser SET status = 3 WHERE cardId = :cardId AND status != 3")
    fun softDeleteByCardId(cardId: Long): CompletableFuture<Void?>

    @Query("DELETE FROM JoinCardWithUser WHERE cardId = :cardId AND userId = :userId")
    fun deletePhysically(cardId: Long, userId: Long): CompletableFuture<Void?>
}
